frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx

```diff
@@ -0,0 +1,221 @@
+import { getCDsPatch } from '../../../k8s/patches/vm/vm-cdrom-patches';
+import { getRemoveDiskPatches } from '../../../k8s/patches/vm/vm-disk-patches';
+import { getVMLikeModel } from '../../../selectors/vm';
+import {
+  getCDRoms,
+  getContainerImageByDisk,
+  getURLSourceByDisk,
+  getPVCSourceByDisk,
+  getStorageSizeByDisk,
+  getStorageClassNameByDisk,
+  isVMRunning,
+} from '../../../selectors/vm/selectors';
+import { VMKind } from '../../../types';
+import { CDRomRow } from './cdrom-row';
+import { initialDisk, WINTOOLS_CONTAINER_NAMES, StorageType } from './constants';
+import './_cdrom-modal.scss';
+
+export const CDRomModal = withHandlePromise((props: CDRomModalProps) => {
+  const {
+    vm,
+    handlePromise,
+    inProgress,
+    errorMessage,
+    persistentVolumeClaims,
+    storageClasses,
+    cancel,
+    close,
+  } = props;
+
+  const getCDsFromResource = (cds) =>
+    cds.map((disk) => {
+      const cd = {
+        ...initialDisk,
+        name: disk.name,
+        bootOrder: disk.bootOrder,
+        isInVM: true,
+      };
+      const container = getContainerImageByDisk(vm, cd.name);
+      if (container) {
+        if (_.includes(WINTOOLS_CONTAINER_NAMES, container))
+          return { ...cd, type: StorageType.WINTOOLS, windowsTools: container };
+        return { ...cd, type: StorageType.CONTAINER, container };
+      }
+
+      const url = getURLSourceByDisk(vm, cd.name);
+      if (url) {
+        const storageClass = getStorageClassNameByDisk(vm, cd.name);
+        const size = getStorageSizeByDisk(vm, cd.name).replace(/[^0-9]/g, '');
+        return { ...cd, type: StorageType.URL, url, storageClass, size };
+      }
+
+      return { ...cd, type: StorageType.PVC, pvc: getPVCSourceByDisk(vm, cd.name) };
+    });
+
+  const [cds, setCDs] = React.useState(getCDsFromResource(getCDRoms(vm)));
+  const [showRestartAlert, setShowRestartAlert] = React.useState(false);
+
+  const onCDChange = (cdName: string, key: string, value: string) => {
+    setCDs(
+      cds.filter((cd) => {
+        if (cdName === cd.name) {
+          cd.changed = true;
+          cd[key] = value;
+          setShowRestartAlert(true);
+
+          if (key === StorageType.URL) {
+            if (!value.startsWith('http://')) cd.isURLValid = false;
+            else cd.isURLValid = true;
+          }
+        }
+        return cd;
+      }),
+    );
+  };
+
+  const onCDAdd = () => {
+    let index = 1;
+    let name = `cd-drive-${index}`;
+    cds.forEach((cd) => {
+      if (cd.name === name) {
+        index++;
+        name = `cd-drive-${index}`;
+      }
+    });
+    const newCD = {
+      ...initialDisk,
+      type: StorageType.CONTAINER,
+      name,
+      isInVM: false,
+      changed: true,
+    };
+    setShowRestartAlert(true);
+    setCDs([...cds, newCD]);
+  };
+
+  const onCDEject = (cdName) =>
+    setCDs(
+      cds.filter((cd) => {
+        if (cdName === cd.name) {
+          cd.ejected = true;
```

Comment:
Let's find/filter first, then change.

Reply:
changed array to object map, no filters anymore